csi/node_server.go

```diff
@@ -487,6 +487,18 @@ func (ns *NodeServer) NodeStageVolume(ctx context.Context, req *csi.NodeStageVol
 		cryptoDevice := crypto.VolumeMapper(volumeID)
 		log.Infof("Volume %s requires crypto device %s", volumeID, cryptoDevice)
 
+		// check if the crypto device is open at the null path.
+		// this will happen if the crypto device is not closed properly and a new attaching request is made on the same node.
+		// reference issue: https://github.com/longhorn/longhorn/issues/9385
+		if isOpenAtNullPath, err := crypto.IsDeviceMappedToNullPath(cryptoDevice); err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		} else if isOpenAtNullPath {
+			log.Infof("Closing active crypto device %s for volume %s", cryptoDevice, volumeID)
+			if err := crypto.CloseVolume(volumeID); err != nil {
```

Comment:
Is it possible that the device is not found? If YES, need to handle it?

Reply:
Modified. Thanks, and `mappedToNullPath` will be `false` and `err` will be `nil` if the device is not found because `mappedFile` is empty returned from the method `DeviceEncryptionStatus`.